uefi/src/proto/network/snp.rs

```diff
@@ -0,0 +1,587 @@
+//! Simple Network Protocol
+//!
+//! Provides a packet level interface to a network adapter.
+//! Once the adapter is initialized, the protocol provides services that allows
+//! packets to be transmitted and received.
+//!
+//! No interface function must be called until `SimpleNetwork.start` is successfully
+//! called first.
+
+use super::{IpAddress, MacAddress};
+use crate::data_types::Event;
+use crate::{Result, Status};
+use core::ffi::c_void;
+use core::ptr;
+use core::ptr::NonNull;
+use uefi_macros::{unsafe_guid, Protocol};
+
+/// The Simple Network Protocol
+#[repr(C)]
+#[unsafe_guid("a19832b9-ac25-11d3-9a2d-0090273fc14d")]
+#[derive(Protocol)]
+pub struct SimpleNetwork {
+    revision: u64,
+    start: extern "efiapi" fn(this: &Self) -> Status,
+    stop: extern "efiapi" fn(this: &Self) -> Status,
+    initialize: extern "efiapi" fn(
+        this: &Self,
+        extra_recv_buffer_size: usize,
+        extra_transmit_buffer_size: usize,
+    ) -> Status,
+    reset: extern "efiapi" fn(this: &Self, extended_verification: bool) -> Status,
+    shutdown: extern "efiapi" fn(this: &Self) -> Status,
+    receive_filters: extern "efiapi" fn(
+        this: &Self,
+        enable: u32,
+        disable: u32,
+        reset_mcast_filter: bool,
+        mcast_filter_count: usize,
+        mcast_filter: Option<NonNull<MacAddress>>,
```

Comment:
nit: Since this can be null I think it would be clearer to write it as `*const MacAddress`.

Reply:
I thought `Option<NonNull<x>>` was preferred to `* x` - at least I saw this in other points of the codebase... Unless you have a strong opinion, I'd leave it like this.

Reply:
We haven't been entirely consistent. I have some thoughts about it (basically, worth doing in some public interfaces but generally not in internal FFI signatures). We have an open bug about deciding how these interfaces should work, but haven't come up with guidelines yet. So, my opinion isn't that strong, feel free to leave as-is.